gh-108277: Add wrapper for timerfd_create, timerfd_settime, and timerfd_gettime to os module (#108382)
python/cpython: Doc/library/os.rst

   .. versionadded:: 3.10

.. function:: timerfd_create(clockid, /, *, flags)

   Create and return a timer file descriptor (*timerfd*).

   The file descriptor returned by :func:`timerfd_create` supports the following

>    additional operations:

Reviewer: ```suggestion
   The file descriptor returned by :func:`timerfd_create` supports:
```

   - :func:`read`
   - :func:`~select.select`
   - :func:`~select.poll`.

   The file descriptor supports :func:`read` with a buffer size of 8.
   If the timer has already expired one or more times, :func:`read` returns
   the number of expirations in :class:`bytes` of host byte order. The return value
   can be coverted to :class:`int` value by :func:`int.from_bytes` with ``byteorder``

>    is :const:`sys.byteorder`.

Reviewer: Thank you for adding more detail! 

What happens if I call ``fd.read()`` and it hasn't expired yet?

```suggestion
   The file descriptor supports :func:`read` with a buffer size of 8.
   If the timer has already expired one or more times, :func:`read` returns
   the number of expirations with the host's endianness,
   which may be coverted to an :class:`int` by ``int.from_bytes(x, byteorder=sys.byteorder)``.
```

This section might be better converted to a recipie in the new subsection I mentioned earlier. Another question: Can I call ``.read()`` multiple times? On a normal file descriptor, at some point I'll reach the end of the file.

Author: @AA-Turner 
Updated.

If there is no timer event after the previous `.read()` call, it blocks until new timer event.

   :func:`~select.select` and :func:`~select.poll` can be used to wait until

>    timer expires and the file descriptor is readable.

Reviewer: Again I think a recipie should be provided

Author: @AA-Turner 
Added examples.

   *clockid* must be a valid :ref:`clock ID <time-clock-id-constants>`,
   as defined in the :py:mod:`time` module and the following values are

>    available.

Reviewer: ```suggestion
   as defined in the :py:mod:`time` module:
```

   specified, :func:`read` blocks.

   Example::

       import time
       import os

       fd = os.timerfd_create(time.CLOCK_REALTIME, 0)

       # timer interval 0.5 second
       inteval = 0.5

       # timer start in 1 second
       value = 1

       # start timer
       _, _ = os.timerfd_settime(fd, 0, inteval, value)

       try:
           while:
               # wait timerfd expired
               _ = fd.read(8)
               do_work()
       finally:
           os.close(fd)

   Example::

       import time
       import os

       fd = os.timerfd_create(time.CLOCK_REALTIME, 0)

       # timer interval 0.5 second
       inteval_ns = 10**9 // 2

       # timer start in 1 second
       value_ns = 10**9

       # start timer in nano second unit
       _, _ = os.timerfd_settime_ns(fd, 0, inteval_ns, value_ns)

       try:
           while:
               # wait timerfd expired
               _ = fd.read(8)
               do_work()
       finally:
           os.close(fd)

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. function:: timerfd_settime(fd, flags, interval, value)

>    Start/stop timer for an timer file descriptor.

Reviewer 2: ```suggestion
   Start/stop timer for a timer file descriptor.
```

>    See man page :manpage:`timerfd_settime(2)` for more information.

Reviewer 2: @AA-Turner, what do you think of using `.. sealso::` for the manpage links?

Author: @erlend-aasland 
Do you mean this?
https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-seealso
@AA-Turner

Reviewer 2: Yes

Author: @erlend-aasland
Fixed.
@AA-Turner

   *interval* coresponds to :c:member:`!it_interval` in :c:struct:`itimerspec` and *value* coresponds to
   :c:member:`!it_value` in :c:struct:`itimerspec`. They are both in seconds and types are double.

   *interval* and *value* are interpreted like this to pass :manpage:`timerfd_settime(2)`.

   .. c:var:: struct itimerspec new_value;
   .. c:var:: double interval;
   .. c:var:: double value;

>    :c:expr:`interval = new_value.it_interval.tv_sec + new_value.it_interval.tv_nsec * 1e-9;`

Reviewer: ```suggestion
   :c:expr:`interval = new_value.it_interval.tv_sec + new_value.it_interval.tv_nsec * 1e-9`
```

Author: @AA-Turner 
Fixed.

>    :c:expr:`value = new_value.it_value.tv_sec + new_value.it_value.tv_nsec * 1e-9;`

Reviewer: ```suggestion
   :c:expr:`value = new_value.it_value.tv_sec + new_value.it_value.tv_nsec * 1e-9`
```

Author: @AA-Turner
Fixed.

   .. versionadded:: 3.10

.. function:: timerfd_create(clockid, flags)

   Create and return a timer file descriptor. The file descriptors supports
   raw :func:`read` with a buffer size of 8, :func:`~select.select`, :func:`~select.poll`

   .. seealso:: :manpage:`timerfd_create(2)`

   Refer to :ref:`time-clock-id-constants` for a list of accepted values for *clockid*.

   *flags* can be constructed from :const:`TFD_NONBLOCK` and :const:`TFD_CLOEXEC`.

   If the timerfd counter is zero and :const:`TFD_NONBLOCK` is not
   specified, :func:`read` blocks.

>    Example::

Reviewer: These two examples feel out-of-place in the documentation for `timerfd_create` specifically. How about adding a new subheading for timer file descriptions (to include all 5 functions / 4 flags) and put the examples in the introduction to this new subsection?

Reviewer: You marked this "resolved", but the examples remain in-line. Is there a comment I missed?

Author: @AA-Turner 
Sorry, I don't remember there is time gap.

> These two examples feel out-of-place in the documentation for timerfd_create specifically. How about adding a new subheading for timer file descriptions (to include all 5 functions / 4 flags) and put the examples in the introduction to this new subsection?

I don't have enough knowledge on rst.
Could you suggest it concretely?

       import time
       import os

       fd = os.timerfd_create(time.CLOCK_REALTIME, 0)

       # timer interval 0.5 second
       inteval = 0.5

       # timer start in 1 second
       value = 1

       # start timer
       _, _ = os.timerfd_settime(fd, 0, inteval, value)

       try:
           while:
               # wait timerfd expired
               _ = fd.read(8)
               do_work()
       finally:
           os.close(fd)

   Example::

       import time
       import os

       fd = os.timerfd_create(time.CLOCK_REALTIME, 0)

       # timer interval 0.5 second
       inteval_ns = 10**9 // 2

       # timer start in 1 second
       value_ns = 10**9

       # start timer in nano second unit
       _, _ = os.timerfd_settime_ns(fd, 0, inteval_ns, value_ns)

       try:
           while:
               # wait timerfd expired
               _ = fd.read(8)
               do_work()
       finally:
           os.close(fd)

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

> .. function:: timerfd_settime(fd, flags, interval, value)

Reviewer: It feels odd to have ``interval`` before ``value`` here -- given ``interval`` can be 0 and the timer still works, but if ``value`` is 0 the timer is turned off.

   Start/stop timer for a timer file descriptor.

   .. seealso:: :manpage:`timerfd_settime(2)`

   *interval* coresponds to :c:member:`!it_interval` in :c:struct:`itimerspec` and *value* coresponds to
   :c:member:`!it_value` in :c:struct:`itimerspec`. They are both in seconds and types are double.

   *interval* and *value* are interpreted like this to pass :manpage:`timerfd_settime(2)`.

   .. c:var:: struct itimerspec new_value;
   .. c:var:: double interval;
   .. c:var:: double value;

   :c:expr:`interval = new_value.it_interval.tv_sec + new_value.it_interval.tv_nsec * 1e-9`

>    :c:expr:`value = new_value.it_value.tv_sec + new_value.it_value.tv_nsec * 1e-9`

Reviewer: This behaviour is surprising. I'd expect that using the nanosecond version of the function overrides any previous values, and vice-versa with the float version of the function -- but it seems their values are combined. I think users should be **warned** of this behaviour.

Author: @AA-Turner 
Added comment at 3d55e2ecff.

Reviewer: I think it is much better to change the suprising behaviour if possible rather than adding a comment. However if you disagree with changing the behaviour, we should rephrase and make this a warning to the user.

```rest
.. caution::
   The seconds set by this function and the nanoseconds set by 
   :func:`timerfd_settime_ns` are summed by the timer descriptor, which may cause surpising results.
   For example, setting an interval of 2 seconds via this function and 2**9 seconds via :func:`timerfd_settime_ns` will make the actual interval of the timer file descriptor **4 seconds**, not 2.
```

Author: @AA-Turner 
What do you mean?
I'm not sure what is your concerns.

Setting an interval of 2 seconds via `timerfd_settime` will make the time file descriptor in every  2 seconds.

Setting an interval of 2**9 seconds(you mean 512 seconds?) via `timerfd_settime_ns` will make the time file descriptor in every 512 seconds.

And the original comment no longer exists on the lastest branch.

   .. versionadded:: 3.10

.. function:: timerfd_create(clockid, flags)

>    Create and return an timerfd file descriptor. The file descriptors supports

Reviewer 3: Should not it be `a timer file descriptor`?

Author: @serhiy-storchaka 
Fixed.

   The file descriptor supports :func:`read` with a buffer size of 8,
   :func:`~select.select`, and :func:`~select.poll`.

>    .. seealso:: The :manpage:`timerfd_create(2)` man page.

Reviewer: The positioning of the ``seealso`` directives is inconsistent -- some are after the intro para and some are at the end.

Author: @AA-Turner 
Moved.

   *clockid* must be a valid :ref:`clock ID <time-clock-id-constants>`,
   as defined in the :py:mod:`time` module.

   The file descriptor's behaviour can be modified by specifying a *flags* value.
   Any of the following variables may used, combined using bitwise OR
   (the ``|`` operator):

   - :const:`TFD_NONBLOCK`
   - :const:`TFD_CLOEXEC`

   If the timer file descriptor's counter is zero and
   :const:`TFD_NONBLOCK` is not set as a flag, :func:`read` blocks.

   The following example shows how to use a timer file descriptor

>    to execute a function twice a second:

Reviewer: Would this run forever (until an exception)? This should be made explicit

Author: @AA-Turner 
Changed the sample code.

               # wait timerfd expired
               _ = fd.read(8)
               do_work()
       finally:
           os.close(fd)

   Example::

       import time
       import os

       fd = os.timerfd_create(time.CLOCK_REALTIME, 0)

       # timer interval 0.5 second
       inteval_ns = 10**9 // 2

       # timer start in 1 second
       value_ns = 10**9

       # start timer in nano second unit
       _, _ = os.timerfd_settime_ns(fd, 0, inteval_ns, value_ns)

       try:
           while:
               # wait timerfd expired
               _ = fd.read(8)
               do_work()
       finally:
           os.close(fd)

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. function:: timerfd_settime(fd, flags, interval, value)

   Start/stop timer for an timerfd file descriptor.
   See man page :manpage:`timerfd_settime(2)` for more information.

   ``interval`` coresponds to ``it_interval`` in ``struct itimerspec`` and ``value`` coresponds to
   ``it_value`` in ``struct itimerspec``. They are both in seconds unit and types are double.

   ``interval`` is calculated like ``it_interval.tv_sec + it_interval.tv_nsec * 1e-9``,

>    ``value`` is calculated like ``it_interval.tv_sec + it_interval.tv_nsec * 1e-9`` internally.

Reviewer 2: The params should be marked up as `*param_name*`, the C struct members with ```:c:member:`!member` ```, and the C structs with ```:c:struct:`!struct s` ```.

I guess we could use `c:expr:` for the calculations, but I'm not sure about the Sphinx magic that happens with `c:expr:`. @AA-Turner would know.

Please also document the return value.

Reviewer: `:c:expr:` adds cross references to everything it can and leaves the rest with literal formatting (similar to using ``` `` ```). `c:texpr:` can be used to render it as normal text instead.

Reference: https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#inline-expressions-and-types

A

Author: @AA-Turner @erlend-aasland 
Fixed.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. function:: timerfd_settime_ns(fd, flags, interval_ns, value_ns)

   Start/stop timer for an timerfd file descriptor.
   See man page :manpage:`timerfd_settime(2)` for more information.

   ``interval_ns`` coresponds to ``it_interval`` in ``struct itimerspec`` and ``value_ns`` coresponds to
   ``it_value`` in ``struct itimerspec``. They are both in nano second unit and types are long long.

   ``interval`` is calculated like ``it_interval.tv_sec * 10**9 + it_interval.tv_nsec``,

>    ``value`` is calculated like ``it_interval.tv_sec * 10**9 + it_interval.tv_nsec`` internally.

Reviewer 2: Ditto

Author: @erlend-aasland 
Fixed.

>    (set by *clockid* in :func:`timerfd_create`) reaches *value* seconds.

Reviewer: Is there any way to query the internal clock given an arbitrary timerfd?

Author: @AA-Turner 
I think there is no such operation.

Reviewer: Should we add one?

Author: @AA-Turner 
You mean we should save the internal clock inside and return the saved value if requested?

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. function:: timerfd_gettime(fd)

>    Returns a tuple ``interval`` which and ``value`` which is the time until next expiration.

Reviewer 2: Please use the imperative mood when writing reference text.

```suggestion
   Return a tuple ``interval`` which and ``value`` which is the time until next expiration.
```

Please make it clearer that the return value is a **two-item** tuple `(interval, value)`.

Reviewer 2: Ditto for `timerfd_gettime_ns`.

Author: @erlend-aasland 
Fixed.

>    the previous timer state, before this function executed.

Reviewer: Same point that the tuple order seems odd

Author: @AA-Turner 
Reorderd.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. function:: timerfd_gettime(fd, /)

>    Return a two-item tuple (``interval``, ``value``).

Reviewer: It's odd that this tuple is (interval, value) -- I'd expect the other way around, as the next expiration is more important than the interval.

   ``value`` denotes the relative time until next the timer next fires,
   regardless of if the :const:`TFD_TIMER_ABSTIME` flag is set.

   ``interval`` denotes the timer's interval.
   If zero, the timer will only fire once, after ``value`` seconds have elapsed.

   .. seealso:: :manpage:`timerfd_gettime(2)`

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. function:: timerfd_gettime_ns(fd, /)

   Similar to :func:`timerfd_gettime`, but return time as nanoseconds.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

> .. data:: TFD_NONBLOCK

Reviewer: All four flags should explain what the *effect* is, not just that they set some attribute at the C level.

Author: @AA-Turner 
Updated.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. data:: TFD_NONBLOCK

   Set :const:`O_NONBLOCK` status flag for new :func:`timerfd_create` file
   descriptor.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. data:: TFD_TIMER_ABSTIME

   Set :const:`TFD_TIMER_ABSTIME` flags for :func:`timerfd_settime` or
   :func:`timerfd_settime_ns`.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. data:: TFD_TIMER_CANCEL_ON_SET

   Set :const:`TFD_TIMER_CANCEL_ON_SET` flags for :func:`timerfd_settime` or
   :func:`timerfd_settime_ns`.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

>    .. versionadded:: 3.13

Reviewer 2: You can collect these constants under one entry. See for example the `SEEK_` or `PRIO_` constants.

Author: @erlend-aasland 
Fixed.

   .. versionadded:: 3.10

> Timer File Descriptors

Author: @AA-Turner 
Thank you for pushing.

~~~~~~~~~~~~~~~~~~~~~~

> .. versionadded:: 3.3

Author: @AA-Turner 
typo for 3.13?

Reviewer: Oops!

   .. versionadded:: 3.10

Timer File Descriptors
~~~~~~~~~~~~~~~~~~~~~~

.. versionadded:: 3.13

These functions provide support for Linux's *timer file descriptor* API.
Naturally, they are all only available on Linux.

The following example shows how to use a timer file descriptor
to execute a function twice a second:

.. code-block:: python

   # Practical scripts should use really use a non-blocking timer,
   # we use a blocking timer here for simplicity.
   import os, time

   # Create the timer file descriptor
   fd = os.timerfd_create(time.CLOCK_REALTIME)

   # Start the timer in 1 second, with an interval of half a second
   os.timerfd_settime(fd, initial=1, interval=0.5)

   try:
       # Process timer events four times.
       for _ in range(4):
          # read() will block until the timer expires
          _ = os.read(fd, 8)

>           print("Timer expired")

Reviewer 3: It looks to me like 3-space indentation instead of PEP 8 recommended 4-space indentation.

Author: @serhiy-storchaka 
I'll fix it.

The indent of `.. code-block:: python` is 3-space.
So I'm confused.

Author: @serhiy-storchaka 
Fixed the indent at f06e3525b3d1ae2f59d0adfba00279d398e3fbd3.

   :func:`~select.select` and :func:`~select.poll` can be used to wait until
   timer expires and the file descriptor is readable.

   *clockid* must be a valid :ref:`clock ID <time-clock-id-constants>`,
   as defined in the :py:mod:`time` module:

   - :const:`time.CLOCK_REALTIME`
   - :const:`time.CLOCK_MONOTONIC`
   - :const:`time.CLOCK_BOOTTIME` (Since Linux 3.15)

   If *clockid* is :const:`time.CLOCK_REALTIME`, a settable system-wide real-time clock
   is used. If system clock is changed, timer setting need to be updated.
   To cancel timer when system clock is changed, see :const:`TFD_TIMER_CANCEL_ON_SET`.

   If *clockid* is :const:`time.CLOCK_MONOTONIC`, A nonsettable monotonically increasing
   clock is used. Even if system clock is changed, timer setting will be not affected.

   If *clockid* is :const:`time.CLOCK_BOOTTIME`, same as :const:`time.CLOCK_MONOTONIC`
   except it includes any time that the system is suspended.

   The file descriptor's behaviour can be modified by specifying a *flags* value.
   Any of the following variables may used, combined using bitwise OR
   (the ``|`` operator):

   - :const:`TFD_NONBLOCK`
   - :const:`TFD_CLOEXEC`

>    If :const:`TFD_NONBLOCK` is not set as a flag, :func:`read` blocks until the timer expires.

Reviewer 3: Too long lines. It makes reviewing difficult. I do not remember the exact limit for documentation, but it is not more than 80 columns.

Author: @serhiy-storchaka 
Fixed.

   .. versionadded:: 3.10

Timer File Descriptors
~~~~~~~~~~~~~~~~~~~~~~

.. versionadded:: 3.13

These functions provide support for Linux's *timer file descriptor* API.
Naturally, they are all only available on Linux.

The following example shows how to use a timer file descriptor
to execute a function twice a second:

.. code-block:: python

   # Practical scripts should use really use a non-blocking timer,
   # we use a blocking timer here for simplicity.
   import os, time

   # Create the timer file descriptor

>    fd = os.timerfd_create(time.CLOCK_REALTIME)

Reviewer 4: I dislike this clock, it can jump backward/forward at sysadmin will. Why using using MONOTONIC for examples?

Author: @vstinner 
1. linux manual uses it as an example.
https://man7.org/linux/man-pages/man2/timerfd_create.2.html

> it can jump backward/forward at sysadmin will

2. Users can detect system clock change by calling timerfd_settime with TFD_TIMER_ABSTIME and TFD_TIMER_CANCEL_ON_SET. If system clock changes, timer will be canceled.
Users need to choose right flag based on the purpose.

Reviewer 4: TFD_TIMER_CANCEL_ON_SET is really nice.

Author: @vstinner 
Resolved?

   # Start the timer in 1 second, with an interval of half a second
   os.timerfd_settime(fd, initial=1, interval=0.5)

   try:
       # Process timer events four times.
       for _ in range(4):
           # read() will block until the timer expires
           _ = os.read(fd, 8)
           print("Timer expired")
   finally:
       # Remember to close the timer file descriptor!
       os.close(fd)

To avoid the precision loss caused by the :class:`float` type,
timer file descriptors allow specifying initial expiration and interval
in integer nanoseconds with ``_ns`` variants of the functions.

This example shows how :func:`~select.epoll` can be used with timer file
descriptors to wait until the file descriptor is ready for reading:

.. code-block:: python

   import os, time, select, socket, sys

   # Create an epoll object
   ep = select.epoll()

   # Create timer file descriptors in non-blocking mode.
   num = 3
   fds = []
   for _ in range(num):
       fd = os.timerfd_create(time.CLOCK_REALTIME, flags=os.TFD_NONBLOCK)
       fds.append(fd)
       # Register the timer file descriptor for read events
       ep.register(fd, select.EPOLLIN)

>    # In this example, use loopback address to send "stop" command to the server.

Reviewer 4: Maybe move it before the creation of the fds, it's more likely to fail if the port is already used.

Author: @vstinner 
Moved.

   #
   # $ telnet 127.0.0.1 1234
   # Trying 127.0.0.1...
   # Connected to 127.0.0.1.
   # Escape character is '^]'.
   # stop
   # Connection closed by foreign host.
   #
   sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
   sock.bind(("127.0.0.1", 1234))
   sock.setblocking(False)
   sock.listen(1)
   ep.register(sock, select.EPOLLIN)

   # Start the timer with os.timerfd_settime_ns() in nanoseconds.
   # Timer 1 fires every 0.25 seconds; timer 2 every 0.5 seconds; etc
   for i, fd in enumerate(fds, start=1):

>        i = i * 10**9

Reviewer 4: 10**9 looks mysterious, please add a constant and give it a name ;-) For example, ONE_SECOND.

Author: @vstinner 
Fixed.

       os.timerfd_settime_ns(fd, initial=i//4, interval=i//4)

   timeout = 3
   try:
       conn = None

>        isActive = True

Reviewer 4: I suggest renaming the variable to is_active.

Author: @vstinner 
Fixed.

   real-time clock is used. If system clock is changed, timer setting need
   to be updated. To cancel timer when system clock is changed, see
   :const:`TFD_TIMER_CANCEL_ON_SET`.

   If *clockid* is :const:`time.CLOCK_MONOTONIC`, a non-settable monotonically
   increasing clock is used. Even if the system clock is changed, the timer
   setting will not be affected.

   If *clockid* is :const:`time.CLOCK_BOOTTIME`, same as :const:`time.CLOCK_MONOTONIC`
   except it includes any time that the system is suspended.

   The file descriptor's behaviour can be modified by specifying a *flags* value.
   Any of the following variables may used, combined using bitwise OR
   (the ``|`` operator):

   - :const:`TFD_NONBLOCK`
   - :const:`TFD_CLOEXEC`

   If :const:`TFD_NONBLOCK` is not set as a flag, :func:`read` blocks until
   the timer expires. If it is set as a flag, :func:`read` doesn't block, but
   If there hasn't been an expiration since the last call to read,
   :func:`read` raises :class:`OSError` with ``errno`` is set to
   :const:`errno.EAGAIN`.

   If :const:`TFD_CLOEXEC` is set as a flag, the close-on-exec flag is set

>    for the new file descriptor.

Reviewer 4: You should explain that it's always set by Python.

Author: @vstinner 
Fixed.

.. function:: timerfd_create(clockid, /, *, flags)

   Create and return a timer file descriptor (*timerfd*).

>    The file descriptor must be closed with :func:`os.close` on cleanup.

Reviewer: I'd move this later in the overview of the function. Also "cleanup" should be clarified -- perhps "to properly clean up the underlying resources" or similar.

Author: @AA-Turner 
Fixed.

>    The file descriptor returned by :func:`timerfd_create` supports:

Reviewer: Could you explain the rationale for splitting this up? I liked the previous version (below), but I'm open to arguments.

Previous version (with endianness update):

```rest
The file descriptor returned by :func:`timerfd_create` supports the following
additional operations:

   - :func:`read`
   - :func:`~select.select`
   - :func:`~select.poll`.

   The file descriptor supports :func:`read` with a buffer size of 8.
   If the timer has already expired one or more times, :func:`read` returns
   the number of expirations with the host's endianness,
   which may be converted to an :class:`int` by ``int.from_bytes(x, byteorder=sys.byteorder)``.

   :func:`~select.select` and :func:`~select.poll` can be used to wait until
   timer expires and the file descriptor is readable.
```

Author: @AA-Turner 
I couldn't find where, but I adapted your suggestion.

Author: @AA-Turner 
Thank you for pushing.

   If the timer file descriptor's counter is zero and
   :const:`TFD_NONBLOCK` is not set as a flag, :func:`read` blocks.

>    If :const:`TFD_CLOEXEC` is set as a flag, set close-on-exec flag for new file descriptor.

Reviewer: Should we explain the close-on-exec flag?

Author: @AA-Turner 
There are some similar consts.
But there are no additional documents.

```
$ git grep CLOEXEC | grep PyModule_AddIntMacro
Modules/fcntlmodule.c:    if (PyModule_AddIntMacro(m, F_DUPFD_CLOEXEC)) return -1;
Modules/fcntlmodule.c:    if (PyModule_AddIntMacro(m, F_DUP2FD_CLOEXEC)) return -1;
Modules/fcntlmodule.c:    if (PyModule_AddIntMacro(m, FD_CLOEXEC)) return -1;
Modules/posixmodule.c:    if (PyModule_AddIntMacro(m, O_CLOEXEC)) return -1;
Modules/posixmodule.c:    if (PyModule_AddIntMacro(m, TFD_CLOEXEC)) return -1;
Modules/posixmodule.c:    if (PyModule_AddIntMacro(m, MFD_CLOEXEC)) return -1;
Modules/posixmodule.c:    if (PyModule_AddIntMacro(m, EFD_CLOEXEC)) return -1;

```

Author: @AA-Turner 
**_CLOEXEC should be documented, 
but I think it's out of the scope of this PR.

   .. seealso:: The :manpage:`timerfd_create(2)` man page.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

> .. function:: timerfd_settime(fd, /, *, flags=flags, initial=0.0, interval=0.0)

Reviewer: ``flags`` isn't a default!
```suggestion
.. function:: timerfd_settime(fd, /, *, flags=0, initial=0.0, interval=0.0)
```

   - :const:`TFD_TIMER_ABSTIME`
   - :const:`TFD_TIMER_CANCEL_ON_SET`

>    The timer is disabled by setting *initial* :py:class:`float` to zero (``0``).

Reviewer: ```suggestion
   The timer is disabled by setting *initial* to zero (``0``).
```

   If *initial* is equal to or greater than zero, the timer is enabled.

>    If *initial* is less than zero, it raises :class:`OSError` with ``errno``
>    set to :const:`errno.EINVAL`

Reviewer: ```suggestion
   If *initial* is less than zero, it raises an :class:`OSError` exception
   with ``errno`` set to :const:`errno.EINVAL`
```

>    By default timer will fire when *initial* seconds have elapsed.

Reviewer: ```suggestion
   By default the timer will fire when *initial* seconds have elapsed.
```

   However, if the :const:`TFD_TIMER_ABSTIME` flag is set,
   the timer will fire when the timer's clock
   (set by *clockid* in :func:`timerfd_create`) reaches *initial* seconds.

   The timer's interval is set by the *interval* :py:class:`float`.
   If *interval* is zero, the timer only fires once, on the initial expiration.
   If *interval* is greater than zero, the timer fires every time *interval*
   seconds have elapsed since the previous expiration.
   If *interval* is less than zero, it raises :class:`OSError` with ``errno``
   set to :const:`errno.EINVAL`

   If the :const:`TFD_TIMER_CANCEL_ON_SET` flag is set along with :const:`TFD_TIMER_ABSTIME`
   and the clock for this timer is :const:`time.CLOCK_REALTIME`, the timer is marked as
   cancelable when if the real-time clock is changed discontinuously. Reading the descriptor
   is aborted with with the error ECANCELED.

   Return a two-item tuple of (``next_expiration``, ``interval``) from
   the previous timer state, before this function executed.

   The following example shows how to use a timer file descriptor
   to execute a function twice a second:

   .. code:: python

      import os, time

>       fd = os.timerfd_create(time.CLOCK_REALTIME, flags=0)

Reviewer: Given flags now defaults to 0:

```suggestion
      fd = os.timerfd_create(time.CLOCK_REALTIME)
```

Author: @AA-Turner 
Updated.

      initial_expiration = 1  # Start the timer in 1 second
      interval = 0.5  # Set the timer interval to 0.5 seconds

>       os.timerfd_settime(fd, flags=0, initial=initial_expiration, interval=interval)  # Start the timer

Reviewer: Similarly:
```suggestion
      os.timerfd_settime(fd, initial=initial_expiration, interval=interval)  # Start the timer
```

Author: @AA-Turner 
Updated.

      try:
          # process timer events four times.
          for _ in range(4):
              fd.read(8)  # Wait for the timer to expire
              do_work()
      finally:
          os.close(fd)

   .. seealso::  The :manpage:`timerfd_settime(2)` man page.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. function:: timerfd_settime_ns(fd, /, *, flags=0, initial=0, interval=0)

   Similar to :func:`timerfd_settime`, but use time as nanoseconds.

   The following example shows how to use a timer file descriptor
   to execute a function twice a second:

   .. code:: python

      import os, time

>       fd = os.timerfd_create(time.CLOCK_REALTIME, flags=0)

Reviewer: ```suggestion
      fd = os.timerfd_create(time.CLOCK_REALTIME)
```

Author: @AA-Turner 
Updated.

      initial_expiration = 10**9  # Start the timer in 1 second
      interval = 10**9 // 2  # Set the timer interval to 0.5 seconds

>       os.timerfd_settime_ns(fd, flags=0, initial=initial_expiration, interval=interval)  # Start the timer

Reviewer: ```suggestion
      os.timerfd_settime_ns(fd, initial=initial_expiration, interval=interval)  # Start the timer
```

Author: @AA-Turner 
When I tried that, I got the following error.
How can you do that?

```
======================================================================
ERROR: test_timerfd_ns_interval (test.test_os.TimerfdTests.test_timerfd_ns_interval)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/..../cpython/Lib/test/test_os.py", line 4096, in test_timerfd_ns_interval
    _, _ = os.timerfd_settime_ns(fd, initial=initial_expiration_ns, interval=interval_ns)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: timerfd_settime_ns() missing required argument 'flags' (pos 2)
```

Author: @AA-Turner 
Updated.

      try:
          # process timer events four times.
          for _ in range(4):
              fd.read(8)  # Wait for the timer to expire
              do_work()
      finally:
          os.close(fd)

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. function:: timerfd_gettime(fd, /)

>    Return a two-item tuple (``next_expiration``, ``interval``).

Reviewer: ```suggestion
   Return a two-item tuple of floats (``next_expiration``, ``interval``).
```

   - :const:`TFD_NONBLOCK`
   - :const:`TFD_CLOEXEC`

   If :const:`TFD_NONBLOCK` is not set as a flag, :func:`read` blocks until the timer expires.
   If it is set as a flag, :func:`read` doesn't block, but if there is no timer expiration,

>    :func:`read` raises :class:`OSError` with ``errno`` is set to :const:`errno.EAGAIN`.

Reviewer: "no timer expiration" -- does this mean if the timer *won't* expire, or if there hasn't been an expiration since the last call to ``read``?

Author: @AA-Turner 
Yes. Exactly.

   .. versionadded:: 3.13

> .. function:: timerfd_settime(fd, flags=0, /, *, initial=0.0, interval=0.0)

Reviewer: Why did we revert to having flags pos-only?
```suggestion
.. function:: timerfd_settime(fd, /, *, flags=0, initial=0.0, interval=0.0)
```

Author: @AA-Turner 
I changed keyword arg, but as I said earlier, I coudn't make it dual arg for pos/keywork.
If you can, please share the concrete way or push the branch directly.

   cancelable when if the real-time clock is changed discontinuously. Reading the descriptor
   is aborted with with the error ECANCELED.

   Linux manages system clock as UTC. A daylight-savings time transition is done

>    by changing time offset only and doesn't cause discontinuous system clock change.

Reviewer: Can we provide an example of what *would* cause a discontinuous system clock change? I'm unsure if we should list something that does not cause a discontinuous change.

Author: @AA-Turner 

https://man7.org/linux/man-pages/man2/timerfd_create.2.html

> if the real-time clock undergoes a discontinuous change
              ([settimeofday(2)](https://man7.org/linux/man-pages/man2/settimeofday.2.html), [clock_settime(2)](https://man7.org/linux/man-pages/man2/clock_settime.2.html), or similar).  When
              such changes occur, a current or future [read(2)](https://man7.org/linux/man-pages/man2/read.2.html) from the
              file descriptor will fail with the error ECANCELED.

   Return a two-item tuple of (``next_expiration``, ``interval``) from
   the previous timer state, before this function executed.

   The following example shows how to use a timer file descriptor
   to execute a function twice a second:

>    Example::

Reviewer: We should prefer ``.. code:: python`` over `::`. Also note that the text above says "the following example", so the "Example:" header isn't needed here

Author: @AA-Turner 
But there is no `.. code:: python` anywhere in the python code tree.
Are you suggesting new comment style here?

To avoid miscommunication, please suggest concrete code.

   A flag for the :func:`timerfd_create` function,
   which sets the :const:`O_NONBLOCK` status flag for the new timer file descriptor.

>    :func:`read` to the new timer file descriptor will not be blocked.

Reviewer: This doesn't read right grammar-wise

Author: @AA-Turner 
Fixed.

.. data:: TFD_TIMER_ABSTIME

   A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions.

>    Use as an absolute value on the timer's clock.

Reviewer: This doesn't make sense

Author: @AA-Turner 
Updated.

   .. availability:: Linux >= 2.6.27 with glibc >= 2.8

   .. versionadded:: 3.13

.. data:: TFD_TIMER_CANCEL_ON_SET

>    A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions
>    along with :const:`TFD_TIMER_ABSTIME`. Timer is aborted when real-time clock is
>    changed discontinuously.

Reviewer: ```suggestion
   A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions
   along with :const:`TFD_TIMER_ABSTIME`.
   The timer is cancelled when the time of the underlying clock changes discontinuously
   (for example, during a daylight-savings time transition).
```

Author: @AA-Turner 
Updated.

   .. versionadded:: 3.13

.. function:: timerfd_settime_ns(fd, flags=0, /, *, initial=0, interval=0)

   Similar to :func:`timerfd_settime`, but use time as nanoseconds.

>    This function operates the same interval timer as :func:`timerfd_settime`.

Reviewer: Same notes as above

Author: @AA-Turner 
Updated.
(https://github.com/python/cpython/pull/108382#discussion_r1334960739)

.. data:: TFD_TIMER_ABSTIME

   A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions.
   If this flag is set, the functions use as an absolute value on the timer's clock

>    in UTC not in local time.

Reviewer: I think this means the same thing:

```suggestion
   If this flag is set, *initial* is interpreted as an absolute value on the timer's clock
   (in UTC seconds or nanoseconds since the Unix Epoch).
```

.. data:: TFD_TIMER_CANCEL_ON_SET

   A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions
   along with :const:`TFD_TIMER_ABSTIME`.

>    The timer is cancelled when the time of the underlying clock changes discontinuously.

Reviewer: This is clearer, thanks! Same note as above -- please could we provide an example of a discontinuous change?

Author: @AA-Turner 
See 
https://github.com/python/cpython/pull/108382#discussion_r1334995659

https://man7.org/linux/man-pages/man2/timerfd_create.2.html

> if the real-time clock undergoes a discontinuous change
              ([settimeofday(2)](https://man7.org/linux/man-pages/man2/settimeofday.2.html), [clock_settime(2)](https://man7.org/linux/man-pages/man2/clock_settime.2.html), or similar).  When
              such changes occur, a current or future [read(2)](https://man7.org/linux/man-pages/man2/read.2.html) from the
              file descriptor will fail with the error ECANCELED.
